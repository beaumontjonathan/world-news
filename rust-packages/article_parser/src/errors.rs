use chrono::ParseError as ChronoParseError;
use htmlescape::DecodeErr;
use scraper::Selector;
use serde_json::Error as SerdeJsonError;

#[derive(Debug)]
pub enum InternalParseError {
  UnknownError(&'static str),
  HtmlStructureError(HtmlStructureError),
  JsonParseError(JsonParseError),
  InvalidDateTime(ChronoParseError),
  HtmlDecodeErr(DecodeErr),
}

#[derive(Debug)]
pub enum JsonParseErrorCause {
  InvalidLdJson,
  InvalidDataAttrJson,
}

#[derive(Debug)]
pub struct JsonParseError {
  pub cause: JsonParseErrorCause,
  pub serde_error: SerdeJsonError,
}

#[derive(Debug)]
pub enum HtmlErrorCause {
  MissingElement,
  NonUniqueElement,
  MissingAttribute(String),
}

#[derive(Debug)]
pub struct HtmlStructureError {
  pub cause: HtmlErrorCause,
  pub selector: &'static Selector,
}

impl InternalParseError {
  pub fn html(error: HtmlStructureError) -> InternalParseError {
    InternalParseError::HtmlStructureError(error)
  }

  pub fn json(error: JsonParseError) -> InternalParseError {
    InternalParseError::JsonParseError(error)
  }

  pub fn ld_json(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidLdJson,
      serde_error,
    })
  }

  pub fn json_data_attr(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidDataAttrJson,
      serde_error,
    })
  }

  pub fn unknown(s: &'static str) -> InternalParseError {
    InternalParseError::UnknownError(s)
  }

  pub fn chrono(err: ChronoParseError) -> InternalParseError {
    InternalParseError::InvalidDateTime(err)
  }

  pub fn decode_html(err: DecodeErr) -> InternalParseError {
    InternalParseError::HtmlDecodeErr(err)
  }
}

impl From<ChronoParseError> for InternalParseError {
  fn from(err: ChronoParseError) -> Self {
    InternalParseError::chrono(err)
  }
}

impl From<DecodeErr> for InternalParseError {
  fn from(err: DecodeErr) -> Self {
    InternalParseError::decode_html(err)
  }
}
