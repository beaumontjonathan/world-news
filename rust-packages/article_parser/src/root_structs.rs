use chrono::{DateTime, Utc};
use serde::Serialize;
use std::fmt::Debug;

#[derive(Debug, Serialize)]
pub struct Image {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Video {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
pub enum Publisher {
  BBC,
  DailyMail,
  Independent,
  Metro,
  Mirror,
  Sky,
  Sun,
  Guardian,
}

use Publisher::*;

impl From<String> for Publisher {
  fn from(publisher: String) -> Self {
    match publisher.to_lowercase().as_str() {
      "bbc" => BBC,
      "dailymail" => DailyMail,
      "independent" => Independent,
      "metro" => Metro,
      "mirror" => Mirror,
      "sky" => Sky,
      "sun" => Sun,
      "theguardian" | "guardian" => Guardian,
      _ => panic!("Invalid publisher type"),
    }
  }
}

#[derive(Debug, Serialize)]
pub struct NewsArticleContent {
  pub headline: String,
  pub twitter_headline: Option<String>,
  pub description: String,
  pub thumbnail: Image,
  pub categories: Vec<String>,
  pub images: Vec<Image>,
  pub videos: Vec<Video>,
  pub body: String,
  pub date_updated: DateTime<Utc>,
  pub date_published: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct VideoArticleContent {
  pub something: String,
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, Serialize)]
pub enum ParsedPageContent {
  NewsArticle(NewsArticleContent),
  VideoArticle(VideoArticleContent),
}

#[derive(Debug, Serialize)]
pub struct PageMeta {
  pub publisher: Publisher,
  pub url: String,
  pub date_parsed: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct ParsedPage {
  pub meta: PageMeta,
  pub content: ParsedPageContent,
}
