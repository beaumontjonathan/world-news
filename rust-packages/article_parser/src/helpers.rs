use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
// use chrono::{DateTime, Utc};
use crate::root_structs::ParsedPageContent;
use lazy_static::lazy_static;
use regex::Regex;
use scraper::{ElementRef, Html, Selector};
// use std::fmt::{self, Debug, Display, Formatter};
pub trait ToOwnedOrNone {
  fn to_owned_or_none(self) -> Option<String>;
}

impl<'a> ToOwnedOrNone for Option<&'a str> {
  fn to_owned_or_none(self) -> Option<String> {
    self.and_then(|mut s| {
      s = s.trim();
      if s.is_empty() {
        None
      } else {
        Some(String::from(s))
      }
    })
  }
}
// pub trait NoneIfEmpty
// where
//   Self: Sized,
// {
//   fn none_if<F>(self, callback: F) -> Option<Self>
//   where
//     F: Fn(&Self) -> bool,
//   {
//     if callback(&self) {
//       None
//     } else {
//       Some(self)
//     }
//   }
// }
// impl<T> NoneIfEmpty for T {}

pub type ParseResult = Result<ParsedPageContent, InternalParseError>;

pub trait FixHtmlWhitespace {
  fn fix_html_whitespace(self) -> Self;
}

fn fix_whitespace(text: String) -> String {
  lazy_static! {
    static ref RE: Regex = Regex::new(r#"\s+"#).unwrap();
  }
  RE.replace_all(&text, " ").into_owned()
}

impl FixHtmlWhitespace for String {
  fn fix_html_whitespace(self) -> String {
    fix_whitespace(self)
  }
}

pub trait AttrValueOrNone {
  fn attr_value_or_none(&self, attr: &str) -> Option<String>;
}

impl<'a> AttrValueOrNone for ElementRef<'a> {
  fn attr_value_or_none(&self, attr: &str) -> Option<String> {
    self.value().attr(attr).to_owned_or_none()
  }
}

pub trait FullFixedHtmlText {
  fn full_fixed_html_text(self) -> String;
}

impl<'a> FullFixedHtmlText for ElementRef<'a> {
  fn full_fixed_html_text(self) -> String {
    self
      .text()
      .map(|text| text.trim().to_string())
      .collect::<Vec<String>>()
      .join(" ")
      .fix_html_whitespace()
      .trim()
      .to_owned()
  }
}

pub trait JoinWithNewline {
  fn join_with_newline(self) -> String;
}

impl<T> JoinWithNewline for T
where
  T: Iterator<Item = String>,
{
  fn join_with_newline(self) -> String {
    self.collect::<Vec<String>>().join("\n").trim().to_owned()
  }
}

pub trait UniqueElementRef<'a> {
  fn unique(self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause>;
}

impl<'a, T> UniqueElementRef<'a> for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn unique(mut self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause> {
    match (self.next(), self.next()) {
      (None, _) => Err(HtmlErrorCause::MissingElement),
      (_, Some(_)) => Err(HtmlErrorCause::NonUniqueElement),
      (Some(el_ref), None) => Ok(el_ref),
    }
  }
}

pub trait SelectHelpers {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError>;
}

impl SelectHelpers for Html {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

impl SelectHelpers for scraper::element_ref::ElementRef<'_> {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

pub trait FormattedText {
  fn formatted_text(self) -> String;
}

impl<'a, T> FormattedText for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn formatted_text(self) -> String {
    self.map(|p| p.full_fixed_html_text()).join_with_newline()
  }
}

pub trait ChildrenHelpers {
  fn children_first(&self) -> Option<ElementRef>;
  fn children_all(&self) -> Vec<ElementRef>;
}

impl ChildrenHelpers for scraper::element_ref::ElementRef<'_> {
  fn children_first(&self) -> Option<ElementRef> {
    self.children_all().into_iter().next()
  }

  fn children_all(&self) -> Vec<ElementRef> {
    self.children().filter_map(ElementRef::wrap).collect()
  }
}
