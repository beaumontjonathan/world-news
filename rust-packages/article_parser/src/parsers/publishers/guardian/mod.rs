use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
use crate::helpers::{
  AttrValueOrNone, ChildrenHelpers, FullFixedHtmlText, JoinWithNewline, ParseResult, SelectHelpers,
};
use crate::root_structs::{Image, NewsArticleContent, ParsedPageContent, Video};

use crate::my_selectors;
use chrono::{DateTime, Utc};
use scraper::{Html, Selector};

my_selectors! {
  ARTICLE_VIDEO_PRECHECK => "body > div > article[id=\"article\"][itemtype=\"http://schema.org/VideoObject\"].content--media--video",
  TWITTER_TITLE_SELECTOR => "head > meta[name=\"twitter:text:title\"]",
  DATE_UPDATED_SELECTOR => "meta[property=\"article:modified_time\"]",
  DATE_PUBLISHED_SELECTOR => "meta[property=\"article:published_time\"]",
  HEADER => "div.content__main-column > header",
  HEADLINE => "h1.content__headline",
  DESCRIPTION => "div.content__standfirst",
  FIGURE_HEADER => "figure.media-primary",
  IMG => "img",
  FIGCAPTION => "figcaption.caption",
  ARTICLE_CONTENT => "div.content__article-body",
  ELEMENT_IMAGE => "figure.element-image",
  ELEMENT_INTERACTIVE => "figure.element-interactive",
  VIDEO_FIGURE => "figure.element[data-atom-type=\"media\"]",
  YOUTUBE_IFRAME => "div.youtube-media-atom__iframe",
}

type Result<T> = std::result::Result<T, InternalParseError>;

pub fn parse(document: Html) -> ParseResult {
  // PRECHECKS
  match document.select(&ARTICLE_VIDEO_PRECHECK).next() {
    None => Ok(()),
    Some(_) => Err(InternalParseError::unknown(
      "Article type: video article, not yet supported",
    )),
    // Some(_) => Err(InternalParseError::precondition(
    //   "Article type: video article, not yet supported",
    // )),
  }?;

  let header = document.select_unique(&HEADER)?;
  let article_content = document.select_unique(&ARTICLE_CONTENT)?;

  // HEADLINE
  let headline = {
    let h1s = header
      .select(&HEADLINE)
      .map(|h1| h1.full_fixed_html_text())
      .collect::<Vec<_>>();
    match h1s.len() {
      0 => Err(InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector: &HEADLINE,
      })),
      1 => Ok(h1s.into_iter().next().unwrap()),
      _ => {
        if h1s.windows(2).all(|w| w[0] == w[1]) {
          Ok(h1s.into_iter().next().unwrap())
        } else {
          Err(InternalParseError::unknown(
            "Unknown headline type - too many <h1>s with differing text",
          ))
        }
      }
    }
  }?;

  // TWITTER HEADLINE
  let twitter_headline = document
    .select_unique_attr(&TWITTER_TITLE_SELECTOR, "content")
    .ok();

  // DESCRIPTION
  let description = header.select_unique(&DESCRIPTION)?.full_fixed_html_text();

  // THUMBNAIL
  // todo - implement
  let thumbnail = {
    let figure = header.select_unique(&FIGURE_HEADER)?;
    let img = figure.select_unique(&IMG)?;
    let alt = img.attr_value_or_none("alt");
    let url = img.attr_value_or_none("src");
    let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
    Image { alt, url, caption }
  };

  // CATEGORIES
  // TODO - implement
  let categories: Vec<String> = vec![];

  // IMAGES
  let images = article_content
    .children_all()
    .into_iter()
    .filter(|el| ELEMENT_IMAGE.matches(&el) || ELEMENT_INTERACTIVE.matches(&el))
    .map(|figure| {
      if ELEMENT_IMAGE.matches(&figure) {
        let img = figure.select_unique(&IMG)?;
        let alt = img.attr_value_or_none("alt");
        let url = img.attr_value_or_none("src");
        let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
        Ok(Image { alt, url, caption })
      } else if ELEMENT_INTERACTIVE.matches(&figure) {
        // unimplemented!();
        Ok(Image {
          alt: None,
          url: None,
          caption: None,
        })
      } else {
        panic!("how are you here?");
      }
    })
    .collect::<Result<Vec<_>>>()?;

  // VIDEOS
  let videos = article_content
    .children_all()
    .into_iter()
    .filter(|figure| VIDEO_FIGURE.matches(&figure))
    .map(|figure| {
      let caption = Some(figure.select_unique(&FIGCAPTION)?.full_fixed_html_text());
      let youtube_id = figure.select_unique_attr(&YOUTUBE_IFRAME, "data-asset-id")?;
      let url = Some(format!("https://www.youtube.com/watch?v={}", youtube_id));

      Ok(Video {
        alt: None,
        url,
        caption,
      })
    })
    .collect::<Result<Vec<_>>>()?;

  // BODY
  let body = article_content
    .children_all()
    .into_iter()
    .filter_map(|el| {
      if el.value().name() == "p" || el.value().name() == "ul" {
        Some(el.full_fixed_html_text())
      } else {
        None
      }
    })
    .collect::<Vec<String>>()
    .into_iter()
    .join_with_newline();

  // DATE_UPDATED
  let date_updated = document.select_unique_attr(&DATE_UPDATED_SELECTOR, "content")?;
  let date_updated = DateTime::parse_from_rfc3339(&date_updated)?.with_timezone(&Utc);

  //DATE_PUBLISHED
  let date_published = document.select_unique_attr(&DATE_PUBLISHED_SELECTOR, "content")?;
  let date_published = DateTime::parse_from_rfc3339(&date_published)?.with_timezone(&Utc);

  Ok(ParsedPageContent::NewsArticle(NewsArticleContent {
    headline,
    twitter_headline,
    description,
    thumbnail,
    categories,
    images,
    videos,
    body,
    date_updated,
    date_published,
  }))
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::test_helpers::helpers::read_html_doc;
  use insta::assert_debug_snapshot;

  #[test]
  fn test_file_01() {
    let document = read_html_doc("test-01.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_02() {
    let document = read_html_doc("test-02.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_03() {
    let document = read_html_doc("test-03.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  // todo - error nicely?
  #[test]
  fn test_file_04() {
    let document = read_html_doc("test-04.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_05() {
    let document = read_html_doc("test-05.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_06() {
    let document = read_html_doc("test-06.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_07() {
    let document = read_html_doc("test-07.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_08() {
    let document = read_html_doc("test-08.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }
}
