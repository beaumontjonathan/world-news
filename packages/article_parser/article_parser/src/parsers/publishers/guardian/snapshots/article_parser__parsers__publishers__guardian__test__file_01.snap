---
source: src/parsers/publishers/guardian/mod.rs
expression: article
---
Ok(
    NewsArticle(
        NewsArticleContent {
            headline: "WH Smith warns on profits as coronavirus empties airport shops",
            twitter_headline: None,
            description: "Trump EU travel ban likely to add to woes as big drop in air travellers slashes revenues Coronavirus - latest news",
            thumbnail: Image {
                alt: Some(
                    "a man passes a WH Smith branch",
                ),
                url: Some(
                    "https://i.guim.co.uk/img/media/4d9433b6fc678850e30fa57846e87b1642f59951/0_25_3000_1800/master/3000.jpg?width=300&quality=85&auto=format&fit=max&s=3df0320f481222fe2e92cf85ca7b6e04",
                ),
                caption: Some(
                    "WH Smith’s UK airport shops account for 60% of travel revenues. Photograph: Alessia Pierdomenico/Reuters",
                ),
            },
            categories: [],
            images: [],
            videos: [],
            body: "WH Smith has become the latest company to issue a profit warning because of the coronavirus pandemic, which has led to a major drop in shoppers at its airport outlets.\nThe retailer’s shares dropped 15% to £13.39 as it warned that the virus outbreak would cost it up to £130m in lost revenues. It said annual profits were likely to be £30m-£40m lower than previously expected. This in effect halves its profit forecast of nearly £80m.\nSome analysts warned the impact could be even worse, after the US president banned most travel from the E U to the country for 30 days. The restrictions will not apply to US citizens or travellers from the UK.\nThe books and stationery store chain flagged a “significant impact” on its shops at airports in the Asia Pacific region, which account for 5% of its travel division’s revenues.\nIn the past fortnight, passenger numbers have dropped significantly at UK airports, where its shops represent 60% of travel revenues; in the US, which accounts for a quarter of travel revenues; and in Europe. WH Smith expects revenues at its UK airport shops to fall by 35% in March and April.\nThe pandemic could also affect high street spending in the UK, the retailer said, and would drag down its full-year revenues and profits.\nNeil Wilson, the chief market analyst at the trading platform Markets.com, said: “You have to feel a little sorry for a company that has done a brilliant job of pivoting away from the struggling British high street to driving all its revenue and profit growth from airports and train stations.\n“A rapid decline in footfall at travel sites because of coronavirus is hitting revenues and will result in a material decline in profits this year. Trump’s 30-day European travel ban makes things worse and threatens to make today’s estimates only partially reflective of the level of damage that could be done this year.”\nIn the six months to 29 February, WH Smith posted a 1% drop in like-for-like revenues. Travel revenues were up 2%, while high street income fell 4%.\nLast May it was rated the UK’s worst high street retailer by consumers for the second year in a row in a Which? survey. Shoppers accused it of having “untidy stores” and offering “poor value for money”.",
            date_updated: 2020-03-18T10:56:06Z,
            date_published: 2020-03-12T13:25:55Z,
        },
    ),
)
